vault/identity_store.go

```diff
@@ -721,34 +721,47 @@ func (i *IdentityStore) invalidateEntityBucket(ctx context.Context, key string)
 				}
 			}
 
-			// If the entity is not in MemDB or if it is but differs from the
-			// state that's in the bucket storage entry, upsert it into MemDB.
-
 			// We've considered the use of github.com/google/go-cmp here,
 			// but opted for sticking with reflect.DeepEqual because go-cmp
 			// is intended for testing and is able to panic in some
 			// situations.
-			if memDBEntity == nil || !reflect.DeepEqual(memDBEntity, bucketEntity) {
-				// The entity is not in MemDB, it's a new entity. Add it to MemDB.
-				err = i.upsertEntityInTxn(ctx, txn, bucketEntity, nil, false)
-				if err != nil {
-					i.logger.Error("failed to update entity in MemDB", "entity_id", bucketEntity.ID, "error", err)
+			if memDBEntity != nil && reflect.DeepEqual(memDBEntity, bucketEntity) {
+				// No changes on this entity, move on to the next one.
+				continue
+			}
+
+			// If the entity exists in MemDB it must differ from the entity in
+			// the storage bucket because of above test. Go through all of the
+			// aliases currently set in the memDB entity and check if any of
+			// those are not in the bucket entity. Those aliases that are only
+			// found in the memDB entity, need to be deleted from MemDB because
+			// the upsertEntityInTxn function does not delete those aliases.
+			if memDBEntity != nil {
```

Comment:
@marcboudreau, doesn't it make more sense to delete the entity altogether from memdb and upsert it? We don't know if the change was only the aliases and we wouldn't like to hit automatic merges during upsert.

Reply:
I mean delete+add it to memdb

Reply:
Also, we need to update the comment above